enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,580 @@
+---
+title: operator-status
+authors:
+  - "@awgreene"
+reviewers:
+  - "@ecordell"
+approvers:
+  - "@kevinrizza"
+creation-date: 2020-05-19
+last-updated: 2020-06-10
+status: implementable
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# operator-status
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to allow operators managed by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) operator to communicate conditions and states that OLM would otherwise be unable to infer. The presence of certain conditions should influence OLM behavior to protect the cluster.
+
+## Motivation
+
+OLM currently infers the state of an operator by inspecting the Kubernetes resources that define the operator. OLM does not provide operators with the means to communicate complex states that cannot be inferred without operator specific knowledge.
+
+Operators managed by OLM would benefit from the ability to communicate their state directly to OLM through a supported channel. Operators could communicate complex conditions and influence how OLM manages the operator, dramatically improving user experience.
+
+A core requirement of this enhancement is to provide operators with the ability to communicate "Upgrade Readiness", providing operators with the ability to signal to OLM that they should not be upgraded at this time. Consider the following scenario based on OLM today:
+
+>An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM could  handle the scenario described above had the operator had the ability to communicate that it should not be upgraded as it ran the critical process.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate a condition that influences OLM behavior, such as the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior if the operator does not take advantage of this feature.
```

Comment:
is there an intent here that OLM will then propagate this status to the CVO in order to block OCP upgrades under certain conditions?

Reply:
Operators saying "don't update me" and "don't update the host cluster where I live" seem like two different things.

Reply:
it's somewhat orthogonal but i bring it up because we are moving forward with documentation that says "make sure all your operators are at the latest version before your upgrade your cluster", so if we know it's physically impossible for your operators to be at the latest version for some reason, it seems worth considering if we should also blocked the cluster upgrade.  (in fact even if it's possible to upgrade your operators but you haven't done so, it would also be nice to block the upgrade).

does not have to be covered here, but wanted it in people's heads.

Reply:
It's not explicit here (because this is an operator-framework enhancement) but the goal here is to allow more communication between OLM <> Operators, and in the future use that to inform the communication between OLM <> CVO and decisions around cluster upgrades. That will be spelled out in a future openshift enhancement, though

Reply:
It might also be worth changing `Upgradeable` to `OperatorUpgradeable` or some such in this enhancement to distinguish between the two cases, so we don't end up with `ClusterUpgradeable` about "please don't bump the host system" and an unqualified `Upgradeable` about "please don't bump me, the operator".

Reply:
>It might also be worth changing Upgradeable to OperatorUpgradeable or some such in this enhancement to distinguish between the two cases, so we don't end up with ClusterUpgradeable about "please don't bump the host system" and an unqualified Upgradeable about "please don't bump me, the operator".

I don't believe that this will be an issue given that this communication channel will exist between OLM and the operator that it manages. OLM would still communicate to CVO using the existing clusteroperator resource.

Reply:
So how would OLM know which sub-operator `Upgradeable=False` we're "hold the core, my current version is incompatible with future core APIs" and should be passed along in the ClusterOperator and which were "don't bump me, my future versions will break on the current cluster" and should not be passed on?

Reply:
The OLM team has discussed the scenarios that @wking is describing but I don't believe that we've met any finite decisions.

That being said, I don't know that operators will be responsible for handling either of these scenarios. They certainly can, but it  places a lot of burden on the operators to understand the requirements of an operator that hasn't been released yet (version n of an operator must know the requirements of operator version n+1). 

In the past, we had discussed having OLM make intelligent decisions when upgrading an operator by looking at available APIs and the RBAC associated with the operator. In this scenario, OLM could identify both of the cases @wking mentioned.

Again, I haven't kept up with this aspect of OLM and would welcome a comment from @ecordell if I am operating off stale info..

Reply:
@wking that's the "other side" of this problem (and isn't something that should be addressed in this operator-framework enhancement). 

This enhancement is just for operators communicating to OLM that they're in the middle of something important, which has been a long-standing problem for upgrades.

With this enhancement done, OLM could later use that to inform CVO that an operator is currently not in an interruptible state (which likely should mean reporting itself `Upgradeable=false` with the ability for an admin to indicate that they don't care about interrupting certain operators). 

But that's not sufficient to determine whether a cluster upgrade is safe. The latest thing we've discussed on that front is that we probably want to _default_ to OLM reporting `Upgradeable=false` to CVO for any operator that's installed that doesn't have an explicit "I work on OCP version range X" where X is the next minor release (the mechanism of how we get that information still under some discussion) - but that will be a new enhancement in openshift/enhancements, since it's OCP-specific.

Reply:
> With this enhancement done, OLM could later use that to inform CVO that an operator is currently not in an interruptible state (which likely should mean reporting itself Upgradeable=false with the ability for an admin to indicate that they don't care about interrupting certain operators).

On the CVO side, `Upgradeable=False` only blocks minor updates, not patch level updates.  If a particular workload wants to mark itself uninterruptible, or otherwise make claims about how interruptible it is, it should use [pod disruption budgets][1].  That will also protect you from things like [deschedulers][2].  So maybe the only thing that needs a new, update-specific channel is "I'm not comfortable with the host environment or other external dependencies updating right now".

[1]: https://kubernetes.io/docs/tasks/run-application/configure-pdb/
[2]: https://github.com/kubernetes-sigs/descheduler